### ISSUE TITLE
TypeError: cafile must be None or a byte string

### ISSUE THREAD

**Issue by wirerydr** [sabnzbd/sabnzbd#223]
sabnzbd:  Git Hash b24ec3e2b7b9a82b9f0b7c2db9e04883a0ea7c30 (branch=develop)
OS:  Fedora 22 x86_64
Python: 2.7.9-6
PyOpenSSL: 0.14-3

When configuring SAB to serve HTTPS with a valid (i.e. non self-signed) cert, I get the following traceback when I specify the chain file I received from my certificate authority:

```
Traceback (most recent call last):
  File "/opt/sabnzbd/cherrypy/process/servers.py", line 187, in _start_http_thread
    self.httpserver.start()
  File "/opt/sabnzbd/cherrypy/wsgiserver/wsgiserver2.py", line 1839, in start
    self.bind(af, socktype, proto)
  File "/opt/sabnzbd/cherrypy/wsgiserver/wsgiserver2.py", line 1892, in bind
    self.socket = self.ssl_adapter.bind(self.socket)
  File "/opt/sabnzbd/cherrypy/wsgiserver/ssl_pyopenssl.py", line 184, in bind
    self.context = self.get_context()
  File "/opt/sabnzbd/cherrypy/wsgiserver/ssl_pyopenssl.py", line 202, in get_context
    c.load_verify_locations(self.certificate_chain)
  File "/usr/lib/python2.7/site-packages/OpenSSL/SSL.py", line 303, in load_verify_locations
    raise TypeError("cafile must be None or a byte string")
TypeError: cafile must be None or a byte string
```

When researching this I discovered [this sabnzbd issue # 209](https://github.com/sabnzbd/sabnzbd/issues/209).  I suspect them to be similar in nature, although unrelated.

I'm not a Python person so I cannot offer a patch that fixes this, and in-fact I'm not sure if there's anything the SAB developers can do since the cause appears to be either in CherryPy, or PyOpenSSL.

However I've been able to make the problem disappear by hacking `/opt/sabnzbd/cherrypy/wsgiserver/ssl_pyopenssl.py` such that this stanza:

```python
    def get_context(self):
        """Return an SSL.Context from self attributes."""
        # See http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/442473
        c = SSL.Context(SSL.SSLv23_METHOD)
        c.use_privatekey_file(self.private_key)
        if self.certificate_chain:
            self.certificate_chain = self.certificate_chain.encode( 'ascii', 'ignore' )
            c.load_verify_locations(self.certificate_chain)
        c.use_certificate_file(self.certificate)
        return c
```

gets changed to:

```python
    def get_context(self):
        """Return an SSL.Context from self attributes."""
        # See http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/442473
        c = SSL.Context(SSL.SSLv23_METHOD)
        c.use_privatekey_file(self.private_key)
        if self.certificate_chain:
##CHANGESTART
            self.certificate_chain = self.certificate_chain.encode( 'ascii', 'ignore' )
##CHANGEEND
            c.load_verify_locations(self.certificate_chain)
        c.use_certificate_file(self.certificate)
        return c
```

I'm almost positive that's not the right way to fix this, and I wouldn't be surprised if this breaks something else, but it does appear to suppress the issue being described.  Firefox subsequently reports my connection as verified by by my CA.

**Comment by shypike** [comment]
As noted by other people (issue 209), PyOpenSSL 0.14 contains some serious bugs in handling Unicode. Please try to upgrade to PyOpenSSL 0.15. I'm not sure we can work around 0.14 bugs without performance degradation for everyone else.

**Comment by shypike** [comment]
Should be solved in 0.8.0Alpha2
